pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -280,9 +293,87 @@ func (r *ReconcileClusterDeployment) Reconcile(request reconcile.Request) (resul
 		return reconcile.Result{}, err
 	}
 
+	rolResult, err := r.reconcileOwnerReferences(cd, cdLog)
+	if err != nil {
+		cdLog.WithError(err).Error("Failed to reconcile objects owned by this cluster deployment.")
+		return reconcile.Result{}, err
+	}
+
+	if rolResult != nil {
+		cdLog.WithError(err).Error("Reconciled objects owned by this cluster deployment.")
+		return *rolResult, nil
+	}
+
 	return r.reconcile(request, cd, cdLog)
 }
 
+func (r *ReconcileClusterDeployment) reconcileOwnerReferences(cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) (*reconcile.Result, error) {
+	selector := map[string]string{constants.OwnerLabel: cd.Name}
+	objects, err := controllerutils.GetRuntimeObjects(r, ownedByClusterDeploymentListTypes, client.InNamespace(cd.Namespace), client.MatchingLabels(selector))
+	if err != nil {
+		errors.Wrap(err, "Failed to list objects owned by ClusterDeployment")
+		return nil, err
+	}
+
+	allErrors := []error{}
+	objectsModified := false
+
+	for _, obj := range objects {
+		tmpModified, err := r.checkAndFixOwnerReferences(cd, obj, cdLog)
+		if tmpModified {
+			objectsModified = true
+		}
+		if err != nil {
+			allErrors = append(allErrors, err)
+		}
+	}
+
+	if len(allErrors) != 0 {
+		return nil, utilerrors.NewAggregate(allErrors)
+	}
+
+	if objectsModified {
+		cdLog.Debug("We successfully reconciled all owner references, and some were fixed.")
+		return &reconcile.Result{
+			RequeueAfter: defaultRequeueTime,
+		}, nil
+	}
+
+	cdLog.Debug("We successfully reconciled all owner references, and none needed to be fixed.")
+	return nil, nil
+}
+
+func (r *ReconcileClusterDeployment) checkAndFixOwnerReferences(cd *hivev1.ClusterDeployment, obj runtime.Object, cdLog log.FieldLogger) (bool, error) {
+	objLog := cdLog.WithField("ownedObject", obj)
+	meta, ok := obj.(metav1.Object)
+	if !ok {
+		return false, fmt.Errorf("Failed converting runtime object to meta object: %v", obj)
+	}
+
+	for _, ownerRef := range meta.GetOwnerReferences() {
+		if *ownerRef.Controller &&
+			ownerRef.APIVersion == cd.APIVersion &&
+			ownerRef.Kind == cd.Kind &&
+			ownerRef.Name == cd.Name {
```

Comment:
Is there a better way to see if an object is owned by another object?

I looked in kube and couldn't find anything that did that (I may have missed it).

Reply:
https://github.com/kubernetes/apimachinery/blob/adc6f4cd9e7d28fc1c1e5efc658d940d55c0f356/pkg/apis/meta/v1/controller_ref.go#L33